### PR TITLE
7926 migrate analytics ga4 class

### PR DESCRIPTION
**Pull request by jimmymadon**
## Summary

Addresses issue:

- #7926

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

**Review by aaemnnosttv — DISMISSED** [review]
Don't merge yet :)

**Review by zutigrm — APPROVED** [review]
Thanks @jimmymadon, LGTM

**Review by nfmohit — CHANGES_REQUESTED** [review]
Incredible hard work here, thank you @jimmymadon!

I have left a few comments.

I'd suggest leaving unused methods intact in the legacy module for now so that we can reference old code easily. There are a few test cases removed from `AnalyticsTest` as well which wouldn't hurt to keep.

Besides, there are a few other minor comments, and `develop` needs to be merged into this branch.

Thank you!

File: includes/Modules/Analytics_4.php [hunk 1]
```diff
@@ -542,17 +549,158 @@ private function create_webdatastream( $property_id, $options = array() ) {
...
+		if ( $this->context->is_amp() ) : ?>
+			<!-- <?php esc_html_e( 'Google Analytics AMP opt-out snippet added by Site Kit', 'google-site-kit' ); ?> -->
+			<meta name="ga-opt-out" content="" id="__gaOptOutExtension">
+			<!-- <?php esc_html_e( 'End Google Analytics AMP opt-out snippet added by Site Kit', 'google-site-kit' ); ?> -->
+		<?php else : ?>
+			<!-- <?php esc_html_e( 'Google Analytics opt-out snippet added by Site Kit', 'google-site-kit' ); ?> -->
+			<?php
+			// Opt-out should always use the measurement ID, even when using a GT tag.
+			$tag_id = $this->get_measurement_id();
+			if ( ! empty( $tag_id ) ) {
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Modules/Analytics_4.php:589]: Ah, yes, that actually makes sense. Let's keep it as it is. Thanks!

File: includes/Modules/Analytics_4.php [hunk 2]
```diff
@@ -557,17 +559,158 @@ private function create_webdatastream( $property_id, $options = array() ) {
...
+			);
+			exit;
+		}
+
+		$new_settings = array();
+
+		// At this point, account creation was successful.
+		$new_settings['accountID'] = $account_id;
+
+		$this->get_settings()->merge( $new_settings );
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:700]: I was just testing this out and found an unfortunate major problem. We're saving the `accountID` here to the `analytics-4` module, however, our current `analytics` settings and overall infrastructure depends on the `accountID` from the `analytics` module, thus, causing an unavailable `accountID` issue.

image

There are 2 probable steps we can take here:

1. In this place, also save the `accountID` for `analytics` and remove in [#7932](https://github.com/google/site-kit-wp/issues/7932).
2. Keep this on hold as well and release it with [#7932](https://github.com/google/site-kit-wp/issues/7932) (more merge conflicts may occur).

Let me know what you think, thanks!

**Review by nfmohit — CHANGES_REQUESTED** [review]
Hi @jimmymadon. I was just testing this out one last time by creating a new account and it looks like we've hit a roadblock. Let me know what you think, thanks!

**Review by nfmohit — APPROVED** [review]
LGTM now, thank you Jimmy! 👍

**Comment by tofumatt** [comment]
We may want to include this change/TODO as well: https://github.com/google/site-kit-wp/blob/3df13cfacbe5cc58d71349d3bc5bab60c4ffc41a/includes/Modules/Analytics_4.php#L639-L640

**Pull request merged**